src/rc.js

```diff
@@ -25,57 +23,76 @@ const buildRcConf = () =>
+// get the built of arguments of a .yarnrc chain of the passed cwd
+function buildRcArgs(cwd: string): Map<string, Array<string>> {
+  const config = getRcConfigForCwd(cwd);
 
-    if (!miniparse) {
-      return argLists;
+  const argsForCommands: Map<string, Array<string>> = new Map();
+
+  for (const key in config) {
+    // args can be prefixed with the command name they're meant for, eg.
+    // `--install.check-files true`
+    const keyMatch = key.match(/^--(?:([^.]+)\.)?(.*)$/);
+    if (!keyMatch) {
+      continue;
     }
 
-    const namespace = miniparse[1] || '*';
-    const arg = miniparse[2];
-    const value = getRcConf()[key];
+    const commandName = keyMatch[1] || '*';
+    const arg = keyMatch[2];
+    const value = config[key];
 
-    if (!argLists[namespace]) {
-      argLists[namespace] = [];
-    }
+    // create args for this command name if we didn't previously have them
+    const args = argsForCommands.get(commandName) || [];
+    argsForCommands.set(commandName, args);
 
+    // turn config value into appropriate cli flag
     if (typeof value === 'string') {
-      argLists[namespace] = argLists[namespace].concat([`--${arg}`, value]);
+      args.push(`--${arg}`, value);
     } else if (value === true) {
-      argLists[namespace] = argLists[namespace].concat([`--${arg}`]);
+      args.push(`--${arg}`);
     } else if (value === false) {
-      argLists[namespace] = argLists[namespace].concat([`--no-${arg}`]);
+      args.push(`--no-${arg}`);
     }
+  }
 
-    return argLists;
-  }, {});
+  return argsForCommands;
+}
 
-export function getRcArgs(command: string): Array<string> {
-  if (!rcArgsCache) {
-    rcArgsCache = buildRcArgs();
+// extract the value of a --cwd arg if present
+function extractCwdArg(args: Array<string>): ?string {
+  for (let i = args.length - 1; i >= 0; i--) {
+    const arg = args[i];
+    if (arg === '--cwd') {
+      return args[i + 1];
+    }
   }
+  return null;
+}
 
-  let result = rcArgsCache['*'] || [];
+// get a list of arguments from .yarnrc that apply to this commandName
+export function getRcArgs(commandName: string, args: Array<string>, previousCwds?: Array<string> = []): Array<string> {
+  // for the cwd, use the --cwd arg if it was passed or else use process.cwd()
+  const origCwd = extractCwdArg(args) || process.cwd();
 
-  if (command !== '*' && Object.prototype.hasOwnProperty.call(rcArgsCache, command)) {
-    result = result.concat(rcArgsCache[command] || []);
-  }
+  // get a map of command names and their arguments
+  const argMap = buildRcArgs(origCwd);
 
-  return result;
-}
+  // concat wildcard arguments and arguments meant for this specific command
+  const newArgs = [].concat(argMap.get('*') || [], argMap.get(commandName) || []);
+
+  // check if the .yarnrc args specified a cwd
+  const newCwd = extractCwdArg(newArgs);
+  if (newCwd && newCwd !== origCwd) {
+    // ensure that we don't enter into a loop
+    if (previousCwds.includes(newCwd)) {
```

Comment:
It's because `String#includes` is implemented in Node 4, but not `Array#includes`